File: leetcode-04/solutions/cpp/0873.cpp
class Solution {
 public:
  int lenLongestFibSubseq(vector<int>& A) {
    const int n = A.size();

    int ans = 0;

    unordered_map<int, int> map;
    for (int i = 0; i < n; i++) map[A[i]] = i;

    vector<vector<int>> dp(n, vector<int>(n, 2));

    for (int j = 0; j < n; j++)
      for (int k = j + 1; k < n; k++)
        if (A[k] - A[j] < A[j] && map.count(A[k] - A[j])) {
          int i = map[A[k] - A[j]];
          dp[j][k] = dp[i][j] + 1;
          ans = max(ans, dp[j][k]);
        }

    return ans;
  }
};

File: leetcode-04/solutions/python3/0873.py
class Solution:
    def lenLongestFibSubseq(self, A: List[int]) -> int:
        n = len(A)

        ans = 0

        dict = {a: i for i, a in enumerate(A)}

        dp = [[2] * n for _ in range(n)]

        for j in range(n):
            for k in range(j + 1, n):
                if A[k] - A[j] < A[j] and A[k] - A[j] in dict:
                    i = dict[A[k] - A[j]]
                    dp[j][k] = dp[i][j] + 1
                    ans = max(ans, dp[j][k])

        return ans
